File: euler05.py
# Какое самое маленькое число делится нацело на все числа от 1 до 20?

from datetime import datetime
from functools import lru_cache


def eratosfen(number):
    """Решето Эратосфена: генерация простых чисел вплоть до заданного числа"""
    sieve = list(range(number + 1))
    sieve[1] = 0
    for i in sieve:
        if i > 1:
            for j in range(2 * i, len(sieve), i):
                sieve[j] = 0
    return sieve


@lru_cache(2 ** 5)
def isprime(number: int):
    """Проверка числа на простоту перебором делителей"""
    if number in {2, 3, 5, 7}: return True
    if number < 2 or number % 2 == 0: return False
    if number % 3 == 0 or number % 5 == 0: return False
    a, k = number, 0
    for i in range(5, int(number**0.5), 6):
        k += 1 if a % i == 0 or a % (i + 2) == 0 else 0
    return True if k <= 0 else False


def euler05():
    n = 20

    # 1 Вариант
    start_time = datetime.now()
    result = 1
    for k in set(eratosfen(n)) - {0}:
        i = 1
        while k ** (i + 1) < n:
            i += 1
        result *= k ** i
    print(result)
    print(datetime.now() - start_time)

    # 2 Вариант
    start_time = datetime.now()
    result = 1
    for k in range(n + 1):
        if isprime(k): result *= k ** max([_ for _ in range(n // 2) if k ** _ < n])
    else:
        print(result)
    print(datetime.now() - start_time)


if __name__ == "__main__":
    euler05()

File: go/euler05.go
// euler05 находит самое маленькое число, делящееся нацело на все числа от 1 до 20.
//
// Решение задачи Эйлера №5.
package main

import (
	"fmt"
	"math"
	"time"
)

// eratosfen генерирует простые числа вплоть до заданного числа (решето Эратосфена).
func eratosfen(number int) []int {
	sieve := make([]int, number+1)
	for i := range sieve {
		sieve[i] = i
	}
	sieve[1] = 0
	for i := 2; i*i <= number; i++ {
		if sieve[i] != 0 {
			for j := 2 * i; j <= number; j += i {
				sieve[j] = 0
			}
		}
	}
	result := make([]int, 0)
	for _, v := range sieve {
		if v != 0 {
			result = append(result, v)
		}
	}
	return result
}

// isPrime проверка числа на простоту перебором делителей.
func isPrime(number int) bool {
	primes := []int{2, 3, 5, 7}
	for _, p := range primes {
		if number == p {
			return true
		}
	}
	if number < 2 || number%2 == 0 {
		return false
	}
	if number%3 == 0 || number%5 == 0 {
		return false
	}
	sqrt := int(math.Sqrt(float64(number)))
	for i := 5; i <= sqrt; i += 6 {
		if number%i == 0 || number%(i+2) == 0 {
			return false
		}
	}
	return true
}

// euler05 решает задачу Эйлера №5.
//
// Какое самое маленькое число делится нацело на все числа от 1 до 20?
func euler05() {
	n := 20
	// Вариант 1
	startTime := time.Now()
	result := 1
	primes := eratosfen(n)
	for _, k := range primes {
		if k > 1 {
			i := 1
			for powInt(k, i+1) < n {
				i++
			}
			result *= powInt(k, i)
		}
	}
	fmt.Println(result)
	fmt.Println(time.Since(startTime))

	// Вариант 2
	startTime = time.Now()
	result = 1
	for k := 2; k <= n; k++ {
		if isPrime(k) {
			maxPower := 1
			for powInt(k, maxPower+1) < n {
				maxPower++
			}
			result *= powInt(k, maxPower)
		}
	}
	fmt.Println(result)
	fmt.Println(time.Since(startTime))
}

func powInt(base, exp int) int {
	result := 1
	for i := 0; i < exp; i++ {
		result *= base
	}
	return result
}

func main() {
	euler05()
}
